mhs-reference-implementation/mhs/SDSLookup/sds.py

```diff
@@ -0,0 +1,105 @@
+
+            self.connection = ldap3.Connection(self.server, auto_bind=True, client_strategy=ldap3.REUSABLE)
+        else:
+            self.server = None
+            self.connection = None
+
+    async def get_mhs_details(self, nhs_id_code: str, service_id: str):
+        """
+        Returns the all end point details for the given org code and service id
+        :param nhs_id_code:
+        :param service_id:
+        :return:
+        """
+
+        accredited_system_lookup = await self._accredited_system_lookup(nhs_id_code, service_id)
+
+        if not accredited_system_lookup:
+            raise Exception('No response from accredited system lookup')
+
+        # As per the spec exactly one result should be returned, manage error checking here
+        response = accredited_system_lookup[0]
+        party_key = response['attributes']['nhsMHSPartyKey']
+
+        if party_key:
+            details = await self._mhs_details_lookup(party_key, service_id)
+            return details[0]
+
+    async def _accredited_system_lookup(self, org_code: str, service_id: str):
+        """
+            Used to find an accredited system, the result contains the ASID and nhsMhsPartyKey.
+            These can then be used to find an MHS endpoint
+        :return: endpoint details - filtered to only contain nhsMHSPartyKey
+        """
+        if not self.connection:
+            raise RuntimeError('No connection established')
+
+        search_filter = f"(&(nhsIDCode={org_code}) (objectClass={AS_OBJECT_CLASS}) (nhsAsSvcIA={service_id}))"
+
+        message_id = self.connection.search(search_base=NHS_SERVICES_BASE,
+                                            search_filter=search_filter,
+                                            attributes="nhsMHSPartyKey")
+
+        task = IOLoop.current().run_in_executor(None, self._get_query_result, message_id)
+        response, result = await task
+        return response
+
+    async def _mhs_details_lookup(self, party_key: str, service_id: str):
+        """
+            Given a party key and a service, this will return an object containing the attributes of that party key,
+            including the endpoint address
+        :return: all the endpoint details
+        """
+        if not self.connection:
+            raise RuntimeError('No connection established')
+
+        search_filter = f"(&(objectClass={MHS_OBJECT_CLASS}) (nhsMhsPartyKey={party_key}) (nhsMhsSvcIA={service_id}))"
+        message_id = self.connection.search(search_base=NHS_SERVICES_BASE,
+                                            search_filter=search_filter,
+                                            attributes=self.mhs_attributes)
+
+        task = IOLoop.current().run_in_executor(None, self._get_query_result, message_id)
+        response, result = await task
+        return response
+
+    def _get_query_result(self, message_id):
+        response, result = self.connection.get_response(message_id)
+        return response, result
```

Comment:
Given we don't seem to care about `result`, is it worth returning it?

Reply:
binned it